base/src/main/java/uk/ac/ebi/atlas/solr/analytics/fullanalytics/FullAnalyticsSearchService.java

```diff
@@ -0,0 +1,182 @@
+                                        expressionThreshold)) :
+                        ImmutableSet.of();
+
+        return
+                mergeGeneProfiles(geneProfilesInSelectedAssayGroups, geneProfilesOutsideSelectedAssayGroups).stream()
+                .sorted(COMPARATOR_BY_SPECIFICITY_AND_NAIVE_DIFF_EXPRESSION)
+                .limit(resultSize)
+                .map(SpecificGeneProfile::geneId)
+                .collect(Collectors.toList());
+    }
+
+    private Set<SpecificGeneProfile> mapJsonBucketsWithAverageExpressionAndCountsToGeneProfiles(String json) {
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+
+        return buckets.stream()
+                .filter(bucket -> bucket.containsKey("avg_expression"))
+                .map(bucket ->
+                        SpecificGeneProfile.create(
+                                (String) bucket.get("val"),
+                                ((Number) bucket.get("count")).intValue(),
+                                ((Number) bucket.get("avg_expression")).doubleValue(),
+                                0, 0.0))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<SpecificGeneProfile> mapJsonBucketsWithMaxExpressionAndCountsToGeneProfiles(String json) {
+        Collection<Map<String, Object>> buckets = JsonPath.read(json, "$.facets.genes.buckets.*");
+
+        return buckets.stream()
+                .filter(bucket -> bucket.containsKey("max_expression"))
+                .map(bucket ->
+                        SpecificGeneProfile.create(
+                                (String) bucket.get("val"),
+                                0, 0.0,
+                                ((Number) bucket.get("count")).intValue(),
+                                ((Number) bucket.get("max_expression")).doubleValue()))
+                .collect(Collectors.toSet());
+    }
+
+    private Set<SpecificGeneProfile> mergeGeneProfiles(
+            Collection<SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSelectionData,
+            Collection<SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSliceData) {
+
+        Map<String, SpecificGeneProfile> geneProfilesInSelectedAssayGroupsWithSliceDataByGeneId =
+                geneProfilesInSelectedAssayGroupsWithSliceData.stream()
+                        .collect(Collectors.toMap(SpecificGeneProfile::geneId, profile -> profile));
+
+        return geneProfilesInSelectedAssayGroupsWithSelectionData.stream()
+                .map(profile ->
+                        SpecificGeneProfile.merge(
+                                profile,
+                                geneProfilesInSelectedAssayGroupsWithSliceDataByGeneId.get(profile.geneId())))
+                .collect(Collectors.toSet());
+    }
+
+    @AutoValue
+    static abstract class SpecificGeneProfile {
```

Comment:
I was very confused by this class, and I didn't understand what it represents, and then I understood that it represents some partial results, and then also the end result you want just before sorting.

Also this class is coupled to AnalyticsDao, and the payload received. I recommend you:
- represent the output of AnalyticsDao by Map<String, Pair<Integer,Double>> (gene id, count, statistic value [max /avg expression])  or something convenient 
- move the logic of creating that to AnalyticsDao
- create `SpecificGeneProfile` from the map entries perhaps, by some extra constructors? 
- remove the merge method from SpecificGeneProfile, and replace with putting together the two maps in `specificExpressionInSliceAcrossSelectedAssayGroups`

Reply:
Very fair point, the coupling is a concern and I went back and forth with several ideas. At one point it was part of the DAO, then it was an independent thing, and it ended as it is. I agree it’s not pretty.